astronomy/orbital_elements_body.hpp

```diff
@@ -0,0 +1,355 @@
+    result.push_back(
+        {/*.t = */ it.time(),
+         /*.a = */ *osculating_elements.semimajor_axis,
+         /*.h = */ e * Sin(ϖ),
+         /*.k = */ e * Cos(ϖ),
+         /*.λ = */ result.empty() ? ϖ + M : UnwindFrom(result.back().λ, ϖ + M),
+         /*.p = */ tg_½i * Sin(Ω),
+         /*.q = */ tg_½i * Cos(Ω),
+         /*.pʹ = */ cotg_½i * Sin(Ω),
+         /*.qʹ = */ cotg_½i * Cos(Ω)});
+  }
+  return result;
+}
+
+inline std::vector<OrbitalElements::EquinoctialElements> const&
+OrbitalElements::osculating_equinoctial_elements() const {
+  return osculating_equinoctial_elements_;
+}
+
+inline std::vector<OrbitalElements::EquinoctialElements> const&
+OrbitalElements::mean_equinoctial_elements() const {
+  return mean_equinoctial_elements_;
+}
+
+inline Time OrbitalElements::SiderealPeriod(
+    std::vector<EquinoctialElements> const& equinoctial_elements) {
+  Time const Δt =
+      equinoctial_elements.back().t - equinoctial_elements.front().t;
+  Instant const t0 = equinoctial_elements.front().t + Δt / 2;
+  Product<Angle, Square<Time>> ſ_λt_dt;
+
+  auto const first = equinoctial_elements.begin();
+  Time const first_t = first->t - t0;
+  Product<Angle, Time> previous_λt = first->λ * first_t;
+  for (auto previous = first, it = first + 1; it != equinoctial_elements.end();
+       previous = it, ++it) {
+    Time const t = it->t - t0;
+    auto const λt = it->λ * t;
+    Time const dt = it->t - previous->t;
+    ſ_λt_dt += (λt + previous_λt) / 2 * dt;
+    previous_λt = λt;
+  }
+  return 2 * π * Radian * Pow<3>(Δt) / (12 * ſ_λt_dt);
+}
+
+inline std::vector<OrbitalElements::EquinoctialElements>
+OrbitalElements::MeanEquinoctialElements(
+    std::vector<EquinoctialElements> const& osculating,
+    Time const& period) {
+  Instant const& t_min = osculating.front().t;
+  struct IntegratedEquinoctialElements {
+    Instant t_max;
+    // The integrals are from t_min to t_max.
+    Product<Length, Time> ſ_a_dt;
+    Time ſ_h_dt;
+    Time ſ_k_dt;
+    Product<Angle, Time> ſ_λ_dt;
+    Time ſ_p_dt;
+    Time ſ_q_dt;
+    Time ſ_pʹ_dt;
+    Time ſ_qʹ_dt;
+  };
+  std::vector<IntegratedEquinoctialElements> integrals;
+  integrals.push_back({t_min});
+  for (auto previous = osculating.begin(), it = osculating.begin() + 1;
+       it != osculating.end();
+       previous = it, ++it) {
+    integrals.push_back(integrals.back());
+    integrals.back().t_max = it->t;
+    Time const dt = it->t - previous->t;
+    integrals.back().ſ_a_dt += (it->a + previous->a) / 2 * dt;
+    integrals.back().ſ_h_dt += (it->h + previous->h) / 2 * dt;
+    integrals.back().ſ_k_dt += (it->k + previous->k) / 2 * dt;
+    integrals.back().ſ_λ_dt += (it->λ + previous->λ) / 2 * dt;
+    integrals.back().ſ_p_dt += (it->p + previous->p) / 2 * dt;
+    integrals.back().ſ_q_dt += (it->q + previous->q) / 2 * dt;
+    integrals.back().ſ_pʹ_dt += (it->pʹ + previous->pʹ) / 2 * dt;
+    integrals.back().ſ_qʹ_dt += (it->qʹ + previous->qʹ) / 2 * dt;
```

Comment:
Should we have a utility class to do a trapezoidal integration?  Would we ever want to change this to something smarter?

Reply:
As long as it is just trapezoidal integration, a utility would probably make things messier: we would have to integrate some sort of functor over some sort of range, which does not sound fun.

We can revisit this if we want to use a fancy quadrature.